test/metrics/test_detection.py

```diff
@@ -0,0 +1,249 @@
+    xyxy=PREDICTIONS[:, :4],
+    confidence=np.ones_like(PREDICTIONS[:, 4]),
+    class_id=PREDICTIONS[:, 5].astype(int),
+)
+
+DETECTION_TENSORS = [
+    np.concatenate(
+        [
+            det.xyxy,
+            np.expand_dims(det.class_id, 1),
+            np.expand_dims(det.confidence, 1),
+        ],
+        axis=1,
+    )
+    for det in [DETECTIONS]
+]
+CERTAIN_DETECTION_TENSORS = [
+    np.concatenate(
+        [det.xyxy, np.expand_dims(det.class_id, 1)],
+        axis=1,
+    )
+    for det in [CERTAIN_DETECTIONS]
+]
+
+IDEAL_MATCHES = np.stack(
+    [
+        np.arange(len(PREDICTIONS)),
+        np.arange(len(PREDICTIONS)),
+        np.ones(len(PREDICTIONS)),
+    ],
+    axis=1,
+)
+
+
+def create_empty_conf_matrix(num_classes: int, do_add_dummy_class: bool = True):
+    if do_add_dummy_class:
+        num_classes += 1
+    return np.zeros((num_classes, num_classes))
+
+
+def update_ideal_conf_matrix(conf_matrix: np.ndarray, class_ids: np.ndarray):
+    for class_id, count in zip(*np.unique(class_ids, return_counts=True)):
+        class_id = int(class_id)
+        conf_matrix[class_id, class_id] += count
+    return conf_matrix
+
+
+IDEAL_CONF_MATRIX = create_empty_conf_matrix(NUM_CLASSES)
+IDEAL_CONF_MATRIX = update_ideal_conf_matrix(IDEAL_CONF_MATRIX, PREDICTIONS[:, 5])
+
+
+DUMMY_DET_DATASET = dummy_detection_dataset_with_map_img_to_annotation()
+DUMMY_DET_IDEAL_CONF_MATRIX = create_empty_conf_matrix(len(DUMMY_DET_DATASET.classes))
+for det in DUMMY_DET_DATASET.annotations.values():
+    DUMMY_DET_IDEAL_CONF_MATRIX = update_ideal_conf_matrix(
+        DUMMY_DET_IDEAL_CONF_MATRIX, det.class_id
+    )
+
+
+@pytest.mark.parametrize(
+    "predictions, targets, classes, conf_threshold, iou_threshold, expected_result, exception",
+    [
+        (
+            [CERTAIN_DETECTIONS],
+            [DETECTIONS],
+            CLASSES,
+            0.3,
+            0.5,
+            IDEAL_CONF_MATRIX,
+            DoesNotRaise(),
+        )
+    ],
+)
+def test_from_detections(
+    predictions,
+    targets,
+    classes,
+    conf_threshold,
+    iou_threshold,
+    expected_result: Optional[np.ndarray],
+    exception: Exception,
+):
+    with exception:
+        result = ConfusionMatrix.from_detections(
+            predictions=predictions,
+            targets=targets,
+            classes=classes,
+            conf_threshold=conf_threshold,
+            iou_threshold=iou_threshold,
+        )
+
+        assert result.matrix.diagonal().sum() == result.matrix.sum()
+        assert np.array_equal(result.matrix, expected_result)
+
+
+@pytest.mark.parametrize(
+    "predictions, targets, classes, conf_threshold, iou_threshold, expected_result, exception",
+    [
+        (
+            DETECTION_TENSORS,
+            CERTAIN_DETECTION_TENSORS,
+            CLASSES,
+            0.3,
+            0.5,
+            IDEAL_CONF_MATRIX,
+            DoesNotRaise(),
+        )
+    ],
+)
+def test_from_tensors(
```

Comment:
I think that now that we have `from_tensors`, we don't have to be shy, and we can use `pytest.mark.parametrize` to provide more edge cases. You no longer need to mock `Detection` and `DetectonDataset`, so it should be waaay easier. 

Let's see what will happen when `predictions` are empty and if `targets` are empty. Let's experiment with different confidence and iou values. Let's see what happens when we have a perfect match `predictions` and `targets` are the same.

I think you don't have to create those const. Just pass `np.array` directly inside `pytest.mark.parametrize`. Same way I done it [here](https://github.com/SkalskiP/onemetric/blob/master/tests/cv/object_detection/test_confusion_matrix.py).

Reply:
I agree on testing from_tensors ruthlessly. But what do we do about test_from_detections() and test_benchmark()? They still require those mocks. Do you suggest removing these two tests?